light-client/src/store/memory.rs

```diff
@@ -81,6 +81,7 @@ impl LightStore for MemoryStore {
             .map(|(_, e)| e.light_block.clone())
     }
 
+    #[allow(clippy::needless_collect)]
```

Comment:
Why can't we just return the boxes iterator directly here? I assume you've tried that already, so just curious what's the issue here.

Reply:
Yeah it's a good question. I refactored the code to the following:

```rust
    fn all(&self, status: Status) -> Box<dyn Iterator<Item = LightBlock>> {
        let light_blocks = self
            .store
            .iter()
            .filter(|(_, e)| e.status == status)
            .map(|(_, e)| e.light_block.clone());

        Box::new(light_blocks)
    }
```

Is that what you mean?

If so, this is the error I get from the compiler:

```
error[E0759]: `self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
  --> light-client/src/store/memory.rs:87:14
   |
84 |       fn all(&self, status: Status) -> Box<dyn Iterator<Item = LightBlock>> {
   |              ----- this data with an anonymous lifetime `'_`...
85 |           let light_blocks = self
   |  ____________________________-
86 | |             .store
   | |__________________- ...is captured here...
87 |               .iter()
   |                ^^^^
...
91 |           Box::new(light_blocks)
   |           ---------------------- ...and is required to live as long as `'static` here
   |
help: to declare that the trait object captures data from argument `self`, you can add an explicit `'_` lifetime bound
   |
84 |     fn all(&self, status: Status) -> Box<dyn Iterator<Item = LightBlock> + '_> {
   |                                                                          ^^^^

error: aborting due to previous error
```

I'm not totally comfortable just yet with the way that these lifetime bounds work (due to lack of experience with using them). Does declaring `that the trait object captures data from argument 'self'` make sense to you? What does this mean, practically?

Reply:
And if I apply the suggestion, we'll need to modify the trait signature:

```
error: `impl` item signature doesn't match `trait` item signature
  --> light-client/src/store/memory.rs:84:5
   |
84 |     fn all(&self, status: Status) -> Box<dyn Iterator<Item = LightBlock> + '_> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ found `fn(&MemoryStore, types::Status) -> Box<dyn Iterator<Item = types::LightBlock>>`
   | 
  ::: light-client/src/store.rs:48:5
   |
48 |     fn all(&self, status: Status) -> Box<dyn Iterator<Item = LightBlock>>;
   |     ---------------------------------------------------------------------- expected `fn(&MemoryStore, types::Status) -> Box<(dyn Iterator<Item = types::LightBlock> + 'static)>`
   |
   = note: expected `fn(&MemoryStore, types::Status) -> Box<(dyn Iterator<Item = types::LightBlock> + 'static)>`
              found `fn(&MemoryStore, types::Status) -> Box<dyn Iterator<Item = types::LightBlock>>`
   = help: the lifetime requirements from the `impl` do not correspond to the requirements in the `trait`
   = help: verify the lifetime relationships in the `trait` and `impl` between the `self` argument, the other inputs and its output
```

Reply:
@thanethomson heh, that's a really "fun" syntax it's nearly impossible to discover without assistance from rustc...

That said, why not write it with `impl Trait` instead of `Box` + `dyn`?

```rust
    fn all(&self, status: Status) -> impl Iterator<Item = LightBlock> + '_ {
       self
            .store
            .iter()
            .filter(|(_, e)| e.status == status)
            .map(|(_, e)| e.light_block.clone())
    }
```

Reply:
@tarcieri Because `all` is a trait method and `impl Trait` return types in trait methods are not supported (yet?)

Reply:
Aah right, would probably require GATs

Reply:
So after refactoring the trait to include the lifetime bound, I get:

```
error[E0373]: closure may outlive the current function, but it borrows `status`, which is owned by the current function
  --> light-client/src/store/memory.rs:88:21
   |
88 |             .filter(|(_, e)| e.status == status)
   |                     ^^^^^^^^             ------ `status` is borrowed here
   |                     |
   |                     may outlive borrowed value `status`
   |
note: closure is returned here
  --> light-client/src/store/memory.rs:91:9
   |
91 |         Box::new(light_blocks)
   |         ^^^^^^^^^^^^^^^^^^^^^^
help: to force the closure to take ownership of `status` (and any other referenced variables), use the `move` keyword
   |
88 |             .filter(move |(_, e)| e.status == status)
   |                     ^^^^^^^^^^^^^

error: aborting due to previous error
```

But the `Status` type implements `Copy`, so why would I need to `move` it into the closure? Is it because the copy would only happen just before the filter closure's executed?

Reply:
Sounds like closures borrow the variables they close over by default, even if they are `Copy`. Moving it into the closure will just copy it into it and should just work(tm)

Reply:
I wonder if we should just return a `Vec` here rather than an iterator. This code is very suboptimal already (and would deserve a rewrite) so I don't think we mind the extra allocations.

Reply:
> I wonder if we should just return a `Vec` here rather than an iterator.

That probably makes sense, given that we're cloning light blocks anyways. So should we perhaps look at changing the method signature to:

```rust
    /// Get an iterator of all light blocks with the given status.
    fn all(&self, status: Status) -> Vec<LightBlock>;
```

Reply:
For now though, perhaps we should just leave the `#[allow(clippy::needless_collect)]`? We can fix this in a follow-up PR, as it touches various aspects of the codebase and is a breaking API change.

Reply:
I've opened #943 to follow up on this.